core/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -188,7 +187,7 @@ public void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexSh
             Map<ShardId, IndexShardSnapshotStatus> shards = snapshotShards.getValue().shards;
             if (shards.containsKey(shardId)) {
                 logger.debug("[{}] shard closing, abort snapshotting for snapshot [{}]", shardId, snapshotShards.getKey().getSnapshotId());
-                shards.get(shardId).abort();
+                shards.get(shardId).moveToAborted("shard is closing, aborting");
```

Comment:
I wonder if we can also use `abortIfNotCompleted` here. Finally, looking at `BlobStoreRepository.snapshot`, we should also do the `store.decRef()` before we move to finalization. The store is unnecessarily kept open (possibly causing ShardLockObtainFailedException) and the snapshot process does not check abortion while it is doing its finalization.

Reply:
> I wonder if we can also use abortIfNotCompleted here.

I think so. I wasn't sure when I added abortIfNotCompleted() but since the abort status is not checked again when finalizing the snapshot we can release the store as soon it is not needed so the shard can be closed and the finalization process continue and succeed or failed as it is already the case today.



> The store is unnecessarily kept open (possibly causing ShardLockObtainFailedException)

Good catch, I changed that too.